### ISSUE TITLE
Async function never returns in Swift 5.7 in release + static linking (Linux)

### ISSUE THREAD

**Issue by PopFlamingo** [apple/swift#59961]
**Describe the bug**
I found a specific project setup where a simple call to a mostly empty async function never returns (note that this doesn't happen in simpler setups without some dependencies, see the [linked repo](https://github.com/PopFlamingo/fluffy-disco) for a reproducible issue).
Also, with some additional dependencies, even if not imported but just listed as dependencies in Package.swift, it can change the behavior and cause a crash with a segfault for instance.

Here, `swift run -c release -Xswiftc -static-executable` never goes beyond printing `"During foo()"`

```swift
func foo() async {
    print("During foo()")
}

@main
struct App {
    static func main() async throws {
        print("Before calling foo()")
        let bar = await foo()
        print("After calling foo()")
    }
}
```

**Steps To Reproduce**
1. Clone `https://github.com/PopFlamingo/fluffy-disco.git`
2. Run it with `swift run -c release -Xswiftc -static-executable` (I can reproduce from the `swiftlang/swift:nightly-5.7-focal` Docker image which currently has the following toolchain `Swift version 5.7-dev (LLVM c94b55955148b74, Swift 13b3ac91a618617)`)

Note that the linked repo has a `.devcontainer` included for easier reproduction in VSCode and others.

**Expected behavior**
Program should print and exit but does not and stops in the `foo()` call.

**Environment (please fill out the following information)**
```
swift --version
Swift version 5.7-dev (LLVM c94b55955148b74, Swift 13b3ac91a618617)
Target: x86_64-unknown-linux-gnu
```

cc @ktoso

**Comment by PopFlamingo** [comment]
I didn't test it on macOS toolchains

**Comment by ktoso** [comment]
Extra note: it does work without the static linking option.

**Comment by ktoso** [comment]
More similar hangs reported over in https://github.com/apple/swift-distributed-actors/issues/1004 in same Linux/docker environment.

**Comment by PopFlamingo** [comment]
I tried it again with the `swift-5.7-DEVELOPMENT-SNAPSHOT-2022-08-02-a` snapshot, there are also some warnings during the build:
```
clang-13: warning: argument unused during compilation: '-pie' [-Wunused-command-line-argument]
/usr/lib/swift_static/linux/libFoundation.a(CFPlatform.c.o):CFPlatform.c:function CFCopyUserName: warning: Using 'getpwuid' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
/usr/lib/swift_static/linux/libFoundation.a(CFPlatform.c.o):CFPlatform.c:function CFCopyUserName: warning: Using 'getpwuid' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
/usr/lib/swift_static/linux/libFoundation.a(CFPlatform.c.o):CFPlatform.c:function CFCopyFullUserName: warning: Using 'getpwuid' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
/usr/lib/swift_static/linux/libFoundation.a(CFPlatform.c.o):CFPlatform.c:function CFCopyFullUserName: warning: Using 'getpwuid' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
/usr/lib/swift_static/linux/libFoundation.a(CFPlatform.c.o):CFPlatform.c:function _CFCopyHomeDirURLForUser: warning: Using 'getpwnam' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
/usr/lib/swift_static/linux/libFoundation.a(CFPlatform.c.o):CFPlatform.c:function _CFCopyHomeDirURLForUser: warning: Using 'getpwuid' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
/usr/lib/swift_static/linux/libFoundation.a(CFBundle_Main.c.o):CFBundle_Main.c:function CFBundleGetMainBundle: warning: Using 'dlopen' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
/usr/lib/swift_static/linux/libFoundation.a(FileManager.swift.o):FileManager.swift.o:function $s10Foundation11FileManagerC17_attributesOfItem6atPath26includingPrivateAttributesSDyAA0B12AttributeKeyVypGSS_SbtKF: warning: Using 'getpwuid' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
/usr/lib/swift_static/linux/libFoundation.a(FileManager.swift.o):FileManager.swift.o:function $s10Foundation11FileManagerC17_attributesOfItem6atPath26includingPrivateAttributesSDyAA0B12AttributeKeyVypGSS_SbtKF: warning: Using 'getgrgid' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
/usr/lib/swift_static/linux/libFoundation.a(CFBundle_Binary.c.o):CFBundle_Binary.c:function _CFBundleDlfcnCheckLoaded: warning: Using 'dlopen' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
/usr/lib/swift_static/linux/libFoundation.a(CFBundle_Binary.c.o):CFBundle_Binary.c:function _CFBundleDlfcnLoadBundle: warning: Using 'dlopen' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
/usr/lib/swift_static/linux/libFoundation.a(CFBundle_Binary.c.o):CFBundle_Binary.c:function _CFBundleDlfcnLoadFramework: warning: Using 'dlopen' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
/usr/lib/swift_static/linux/libFoundation.a(Host.swift.o):Host.swift.o:function $s10Foundation4HostC8_resolveyyF: warning: Using 'getaddrinfo' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
[12/12] Linking Run
```

**Comment by ktoso** [comment]
I believe this wasn't an actual issue, since in the other bug @PopFlamingo confirmed that it was a logging issue instead:

> It turns out this specific issue was actually logging related. Since I am not getting the logs directly from a shell, stdout needed to be flushed as Swift doesn't automatically do so after a print.
Thanks for your help! I will close this now.

**Comment by PopFlamingo** [comment]
@ktoso I checked again and this is still an actual issue. I tested it for the first time on macOS Docker today and I get the same hang as on Linux Docker.

**Comment by ktoso** [comment]
Thanks for confirming, that's no good then :-\

**Comment by Austinpayne** [comment]
Experimentally, this appears to be resolved on the latest `swift:5.9-jammy` image (as of today).

**Comment by hborla** [comment]
I believe this was resolved by https://github.com/apple/swift-corelibs-foundation/pull/4672
